### PR TITLE
Correct hydration of resultsets coming from DQL queries where the root of the selection is aliased and also the root of an inheritance

### PR DESCRIPTION
**Pull request by SirWaddles**
Resolves Issue #6362 

Added a test case for reproducability.

### CODE REVIEW

**Comment by SirWaddles** [comment]
Travis failures are unrelated, something to do with the Symfony Yaml component.

**Comment by SirWaddles** [comment]
The main reason for test is to check that the Undefined Property error does not occur, and if you run the test without the fix, it will error out. It does however also make some assertions about the format of the result which may be questionable.

The resultset only has one row, yet the hydrated result has two. My assumption was that this is intentional as this change did not break any tests, but it does seem odd to me.

**Comment by SirWaddles** [comment]
Might be worth running the tests again, only one job failed and it had just stalled before even getting to my code. I have no idea why my change would cause only PHP7.1/MariaDB to halt.

**Comment by Ocramius** [comment]
@SirWaddles restarted

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH6362Test.php [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                $this->_em->getClassMetadata(GH6362Start::class),
+                $this->_em->getClassMetadata(GH6362Base::class),
+                $this->_em->getClassMetadata(GH6362Child::class),
+                $this->_em->getClassMetadata(GH6362Join::class),
+            ]
+        );
+    }
+
+    /**
+     * SELECT a as base, b, c, d
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH6362Test.php]: `@group GH6362` missing

**Review by lcobucci — COMMENTED** [review]
The rest LGTM, what do you think @Ocramius?

**Comment by lcobucci** [comment]
@SirWaddles I've rebased your branch to sync with the latest updates and group the commits. Patch LGTM but I'd love to have @guilhermeblanco's and/or @Ocramius' opinion just to double check.

Thanks for the contribution 😉

**Comment by lcobucci** [comment]
Quite easy to backport for `v2.5.7` too...

**Review by Ocramius — APPROVED** [review]

**Comment by Ocramius** [comment]
Good to go! Will also backport :+1: 

Thanks @SirWaddles, pristine work.

**Comment by Ocramius** [comment]
Hah, I backported to `2.5` via doctrine/doctrine2@04fc7a9a1ccd377c33ec753151b425fc3db22c27, but it will obviously fail due to lack of PHP 5.4 compliance =_=

**Comment by Ocramius** [comment]
Should be corrected in doctrine/doctrine2@d2c805b07120e0d31cd4c27dbe39124ccec65092 (see https://travis-ci.org/doctrine/doctrine2/builds/227851219)

**Pull request merged**